File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.java
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> stack=new Stack<>();
        Set<Character> bools=new HashSet<>();
        bools.add('t');bools.add('f');
        Set<Character> operators=new HashSet<>();
        operators.add('!');operators.add('|');operators.add('&');
        for(char s:expression.toCharArray()){
            if(operators.contains(s) || bools.contains(s)){
                stack.push(s);
            }
            else if(s==')'){
                Boolean containsTrue=false;
                Boolean containsFalse=false;
                while(!operators.contains(stack.peek())){
                    char top=stack.pop();
                    if(top=='t'){
                        containsTrue=true;
                    }
                    else if(top=='f'){
                        containsFalse=true;
                    }
                }
                char operatorForEvaluvation=stack.pop();
                if(operatorForEvaluvation=='!'){
                    stack.push(containsTrue ? 'f':'t');
                }
                else if(operatorForEvaluvation=='&'){
                    stack.push(containsFalse ? 'f':'t');
                }
                else{
                    stack.push(containsTrue ? 't':'f');
                }
            }
        }
        return stack.pop()=='t';
    }
}

File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stack=[]
        operators={'|','&','!'}
        bools={'t','f'}
        for s in expression:
            if s in operators or s in bools:
                stack.append(s)
            elif s==')':
                contains_true,contains_false=False,False
                while stack[-1] not in operators:
                    top=stack.pop()
                    if top=='t':
                        contains_true=True
                    elif top=='f':
                        contains_false=True
                operator_eval=stack.pop()
                if operator_eval=='!':
                    if contains_true:
                        stack.append('f')
                    elif contains_false:
                        stack.append('t')
                elif operator_eval=='&':
                    if contains_false:
                        stack.append('f')
                    else:
                        stack.append('t')
                else:
                    if contains_true:
                        stack.append('t')
                    else:
                        stack.append('f')
        return stack[-1]=='t'
                
    

        